In [1]:


import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import numpy as np
import matplotlib.pyplot as plt
import os


## Part 1. Data Analysis and Bayes Nets.  

### 1.1 standard data set标准数据集

In [2]:


X = np.load('/home/mw/input/smiley2472/smiley_X.npy')
y = np.load('/home/mw/input/smiley2472/smiley_Y.npy')
# pd.DataFrame(X)
X.shape,y.shape


((204, 9, 9, 1), (204,))

In [3]:
import matplotlib.pyplot as plt

# 创建子图create subgraph
fig, axes = plt.subplots(3,3)


# 循环遍历图片文件，并在每个子图上显示对应的图片Loop through the image file and display the corresponding image on each subgraph
for i, ax in enumerate(axes.flat):
    # 读取图片文件并显示Read image file and display
#     image = plt.imread(image_files[i])
    ax.imshow(X[i+20,:,:,0])
    
    # 设置子图标题Set the subplot title
    ax.set_title(f"Image {i+1}")
    # 添加副标题add subtitle
    ax.text(0.5, -0.5, y[i+20], fontsize=10, ha='center', transform=ax.transAxes)
    
    # 隐藏坐标轴Hide Axes
    ax.axis('off')

# 调整子图之间的间距和布局Adjust spacing and layout between subplots
plt.tight_layout()

# 显示图形display graphics
plt.show()


<Figure size 432x288 with 9 Axes>

### 1.2 noise dataset噪音数据集

In [4]:
X1 = np.load('/home/mw/input/smiley2472/smiley_noisy_X.npy')
y1 = np.load('/home/mw/input/smiley2472/smiley_noisy_Y.npy')
# pd.DataFrame(X)
X1.shape,y1.shape
# y1

((204, 9, 9, 1), (204,))

In [5]:
# plt.imshow(X[50,:,:,0])
# pd.DataFrame(X[50,:,:,0])

# plt.imshow(X1[50,:,:,0])
# pd.DataFrame(X1[50,:,:,0])

import matplotlib.pyplot as plt

# 创建子图create subgraph
fig, axes = plt.subplots(3,3)


# 循环遍历图片文件，并在每个子图上显示对应的图片Loop through the image file and display the corresponding image on each subgraph
for i, ax in enumerate(axes.flat):
    # 读取图片文件并显示Read image file and display
    ax.imshow(X1[i+20,:,:,0])
    
    # 设置子图标题Set the subplot title
    ax.set_title(f"Image {i+1}")
    # 添加副标题add subtitle
    ax.text(0.5, -0.5, y1[i+20], fontsize=10, ha='center', transform=ax.transAxes)
    
    # 隐藏坐标轴Hide Axes
    ax.axis('off')

# 调整子图之间的间距和布局Adjust spacing and layout between subplots
plt.tight_layout()

# 显示图形display graphics
plt.show()


<Figure size 432x288 with 9 Axes>

### 1.3 Filtering滤波处理

在图像识别机器学习任务中，滤波处理是一个常用的预处理步骤，以下是一些滤波处理的一些优点：  

- **去除噪声**：图像可能受到各种类型的噪声污染，例如高斯噪声、椒盐噪声等。这些噪声会干扰对图像信息的正确理解和分析。通过应用适当的滤波器，可以降低或去除这些噪声，提高图像质量和后续处理的准确性。  
- **平滑图像**：有时候，图像中的细节过多或过于突出可能对后续处理任务造成困扰。使用平滑滤波器可以模糊图像，减少细节，使得图像更加平滑和均匀，以便更好地处理和分析。  
- **特征增强**：某些情况下，我们希望突出或增强图像中的某些特定特征，例如边缘、纹理等。通过应用合适的滤波器，可以增强这些特征，使得它们更加明显和易于识别。  

需要注意的是，滤波处理应根据具体的任务和图像特点进行选择和调整。不同的滤波器和参数会产生不同的效果，需要根据实际情况进行优化和调试。  

滤波处理可以改善**图像质量、增强特征、去除噪声**等，从而为后续的图像识别机器学习任务提供更好的输入数据。  
In image recognition machine learning tasks, filtering is a common preprocessing step. Here are some advantages of filtering:  

- **Noise removal**: The image may be polluted by various types of noise, such as Gaussian noise, salt and pepper noise, etc. These noises can interfere with proper understanding and analysis of image information. By applying appropriate filters, these noises can be reduced or removed, improving image quality and subsequent processing accuracy.  
- **Smooth Image**: Sometimes too much or too much detail in an image can be a problem for subsequent processing tasks. Using a smoothing filter can blur the image, reduce details, and make the image smoother and more uniform for better processing and analysis.  
- **Feature Enhancement**: In some cases, we want to highlight or enhance some specific features in the image, such as edges, textures, etc. By applying suitable filters, these features can be enhanced to make them more obvious and recognizable.  

It should be noted that the filtering process should be selected and adjusted according to specific tasks and image characteristics. Different filters and parameters will produce different effects, which need to be optimized and debugged according to the actual situation.  

Filtering can improve** image quality, enhance features, remove noise**, etc., thereby providing better input data for subsequent image recognition machine learning tasks.

#### **中值滤波去噪**  

中值滤波：中值滤波是一种非常简单但有效的去噪方法。它通过将每个像素的值替换为它周围像素值的中值来去除噪声。  

可以使用OpenCV库中的cv2.medianBlur函数来应用中值滤波。  
#### **Median filter denoising**  

Median filtering: Median filtering is a very simple but effective method of denoising. It removes noise by replacing the value of each pixel with the median value of the pixel values around it.  

Median filtering can be applied using the cv2.medianBlur function in the OpenCV library.

In [6]:
# ! pip install cv2
!pip install -i https://pypi.tuna.tsinghua.edu.cn/simple opencv-python

Looking in indexes: https://pypi.tuna.tsinghua.edu.cn/simple


In [7]:
import matplotlib.pyplot as plt
import cv2
# 创建子图create subgraph
fig, axes = plt.subplots(3, 3)


# 循环遍历图片文件，并在每个子图上显示对应的图片Loop through the image file and display the corresponding image on each subgraph
for i, ax in enumerate(axes.flat):
    # 读取图片文件并显示Read image file and display
    
    image=X1[i+20,:,:,0]
    # 将图像数据类型转换为8位无符号整数类型Convert image data type to 8-bit unsigned integer type
    image = cv2.convertScaleAbs(image)
    # 应用双边滤波apply bilateral filtering
#     image.astype(np.uint8)
    
#     denoised_image = cv2.bilateralFilter(image,5, 75, 75)

#     # 应用中值滤波apply median filter
    denoised_image = cv2.medianBlur(image,7)
    
#     # 应用高斯滤波apply Gaussian filtering
#     denoised_image = cv2.GaussianBlur(X1[i+20,:,:,0], (5, 5), 0)
    ax.imshow(denoised_image)
    
    # 设置子图标题Set the subplot title
    ax.set_title(f"Image {i+1}")
    # 添加副标题add subtitle
    ax.text(0.5, -0.5, y1[i+20], fontsize=10, ha='center', transform=ax.transAxes)
    
    # 隐藏坐标轴Hide Axes
    ax.axis('off')

# 调整子图之间的间距和布局Adjust spacing and layout between subplots
plt.tight_layout()

# 显示图形display graphics
plt.show()

<Figure size 432x288 with 9 Axes>

In [8]:
denoised_image

array([[0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=uint8)

可以看到，滤波之后，图像的值全部为0，这说明**均值滤波算法不适用于该数据集**。  
It can be seen that after filtering, the values of the image are all 0, which shows that the mean filtering algorithm is not suitable for this data set.

####  **高斯滤波去噪**  

**高斯滤波**：高斯滤波是一种常用的线性平滑滤波器，可以有效地去除高斯噪声。  

它通过对每个像素周围的像素进行加权平均来平滑图像。  

在Python中，可以使用OpenCV库中的 **cv2.GaussianBlur** 函数来应用高斯滤波。  
Gaussian filter denoising  
Gaussian filtering: Gaussian filtering is a commonly used linear smoothing filter that can effectively remove Gaussian noise.  

It smoothes the image by taking a weighted average of the pixels surrounding each pixel.  

In Python, Gaussian filtering can be applied using the cv2.GaussianBlur function from the OpenCV library.  


In [9]:
import matplotlib.pyplot as plt

# 创建子图create subgraph
fig, axes = plt.subplots(3, 3)


# 循环遍历图片文件，并在每个子图上显示对应的图片Loop through the image file and display the corresponding image on each subgraph
for i, ax in enumerate(axes.flat):
    # 读取图片文件并显示Read image file and display
    
    # 应用高斯滤波apply Gaussian filtering
    denoised_image = cv2.GaussianBlur(X1[i+1,:,:,0], (5, 5), 0)
    ax.imshow(denoised_image)
    
    # 设置子图标题Set the subplot title
    ax.set_title(f"Image {i+1}")
    # 添加副标题add subtitle
    ax.text(0.5, -0.5, y1[i+1], fontsize=10, ha='center', transform=ax.transAxes)
    
    # 隐藏坐标轴Hide Axes
    ax.axis('off')

# 调整子图之间的间距和布局Adjust spacing and layout between subplots
plt.tight_layout()

# 显示图形display graphics
plt.show()

<Figure size 432x288 with 9 Axes>

In [10]:
denoised_image

array([[0.0625    , 0.0546875 , 0.03125   , 0.0078125 , 0.        ,
        0.0078125 , 0.03125   , 0.0546875 , 0.0625    ],
       [0.125     , 0.109375  , 0.0625    , 0.015625  , 0.        ,
        0.015625  , 0.0625    , 0.109375  , 0.125     ],
       [0.21875   , 0.19140625, 0.109375  , 0.02734375, 0.        ,
        0.02734375, 0.109375  , 0.19140625, 0.21875   ],
       [0.2578125 , 0.23828125, 0.16796875, 0.08984375, 0.0625    ,
        0.08984375, 0.16796875, 0.23828125, 0.2578125 ],
       [0.25      , 0.26953125, 0.28125   , 0.26171875, 0.25      ,
        0.26171875, 0.28125   , 0.26953125, 0.25      ],
       [0.171875  , 0.2265625 , 0.3203125 , 0.3671875 , 0.375     ,
        0.3671875 , 0.3203125 , 0.2265625 , 0.171875  ],
       [0.0625    , 0.10546875, 0.1875    , 0.23828125, 0.25      ,
        0.23828125, 0.1875    , 0.10546875, 0.0625    ],
       [0.0078125 , 0.01953125, 0.04296875, 0.05859375, 0.0625    ,
        0.05859375, 0.04296875, 0.01953125, 0.0078125 ],


**可以看到，高斯滤波是有效的，能够对图像核心点周围的数据进行平滑处理。**  
It can be seen that Gaussian filtering is effective and can smooth the data around the core point of the image.

### 1.4 Naïve Bayes Classifier朴素贝叶斯分类器

 #### 1.4.1  **正常数据**normal data

**(1) Naive Bayes**

In [11]:
import warnings 
warnings.filterwarnings('ignore') 
print(X.shape,"\n",X[0,:,:,0])
X[0,:,:,0].reshape(-1,X.shape[1]*X.shape[1])

X_changed=pd.DataFrame(X[0,:,:,0].reshape(-1,X.shape[1]*X.shape[1]))

for i in range(1,X.shape[0]):
    tmp = pd.DataFrame(X[i,:,:,0].reshape(-1,X.shape[1]*X.shape[1]))
    #将 df2 的行追加到现有 DataFrame 的末尾append rows of df2 to end of existing DataFrame
#     X_changed.loc[i]=tmp
    X_changed = X_changed.append(tmp)
    X_changed = X_changed.reset_index().drop(["index"],axis=1)
#     X_changed=X_changed.reset_index()


X_changed["label"]=y
df_changed = X_changed.copy()
df_changed

(204, 9, 9, 1) 
 [[0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 1. 0. 0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 1. 0.]
 [0. 0. 1. 0. 0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 1. 1. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0.]]


0    1    2    3    4    5    6    7    8    9  ...   72   73   74  \
0    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
3    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
4    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
5    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
6    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
7    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
8    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
9    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
10   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
11   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
12   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
13   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
14   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
15   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
16   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
17   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
18   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
19   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
20   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
21   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
22   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
23   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
24   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
25   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
26   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
27   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
28   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
29   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
174  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
175  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
176  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
177  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
178  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
179  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  1.0  0.0   
180  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
181  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  1.0   
182  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  1.0   
183  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
184  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
185  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
186  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
187  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  1.0  0.0  0.0   
188  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
189  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
190  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
191  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
192  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
193  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
194

In [12]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report, roc_curve, auc
import numpy as np
import matplotlib.pyplot as plt
from itertools import combinations
# 引入必要的库
import numpy as np
import matplotlib.pyplot as plt
from itertools import cycle
from sklearn import svm, datasets
from sklearn.metrics import roc_curve, auc
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import label_binarize
from sklearn.multiclass import OneVsRestClassifier
from scipy import interp
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report, confusion_matrix, roc_curve, auc
import numpy as np
import matplotlib.pyplot as plt



X = df_changed.drop(["label"],axis=1)
y = df_changed.label


# 洗牌和分割训练集和测试集shuffle and split training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.2,random_state=0)

# 创建朴素贝叶斯分类器Create a Naive Bayes Classifier
clf = GaussianNB()

# 拟合模型fit model
clf.fit(X_train, y_train)

# 在测试集上进行预测Make predictions on the test set
y_pred = clf.predict(X_test)

# 计算混淆矩阵Calculate the confusion matrix
cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(cm)

# 计算分类报告Calculate Classification Report
report = classification_report(y_test, y_pred)
print("Classification Report:")
print(report)




Confusion Matrix:
[[12  3  0]
 [ 0  8  1]
 [ 0  7 10]]
Classification Report:
              precision    recall  f1-score   support

         0.0       1.00      0.80      0.89        15
         1.0       0.44      0.89      0.59         9
         2.0       0.91      0.59      0.71        17

    accuracy                           0.73        41
   macro avg       0.78      0.76      0.73        41
weighted avg       0.84      0.73      0.75        41



In [13]:
# 在测试集上进行预测Make predictions on the test set
y_pred = clf.predict_proba(X_test)
y_pred

classes=[0,1,2]

y_test_all = label_binarize(y_test, classes=[0,1,2])

# 在测试集上进行预测Make predictions on the test set
y_pred = clf.predict_proba(X_test)
y_pred

y_score_all=y_pred
fpr = dict()
tpr = dict()
roc_auc = dict()
for i in range(len(classes)):
    fpr[i], tpr[i], thresholds = roc_curve(y_test_all[:, i],y_score_all[:, i])
    roc_auc[i] = auc(fpr[i], tpr[i])

roc_auc

# micro-average ROC curve（方法一）
fpr["micro"], tpr["micro"], thresholds = roc_curve(y_test_all.ravel(),y_pred.ravel())
roc_auc["micro"] = auc(fpr["micro"], tpr["micro"])
roc_auc



# macro-average ROC curve 方法二）

all_fpr = np.unique(np.concatenate([fpr[i] for i in range(len(classes))]))

mean_tpr = np.zeros_like(all_fpr)
for i in range(len(classes)):
    mean_tpr += interp(all_fpr, fpr[i], tpr[i])
# 求平均计算ROC包围的面积AUC(Calculate the average area AUC surrounded by ROC
mean_tpr /= len(classes)
fpr["macro"] = all_fpr
tpr["macro"] = mean_tpr
roc_auc["macro"] = auc(fpr["macro"], tpr["macro"])

#画图部分drawing part
plt.figure()
plt.plot(fpr["micro"], tpr["micro"],'k-',color='y',
         label='ROC curve micro-average(AUC = {0:0.4f})'
               ''.format(roc_auc["micro"]),
          linestyle='-.', linewidth=3)

plt.plot(fpr["macro"], tpr["macro"],'k-',color='k',
         label='ROC curve macro-average(AUC = {0:0.4f})'
               ''.format(roc_auc["macro"]),
          linestyle='-.', linewidth=3)
plt.plot([0,1],[0,1],'r--')
plt.xlim([-0.1,1.1])
plt.ylim([-0.1,1.1])
plt.ylabel('True Positive Rate')
plt.xlabel('False Positive Rate')
plt.legend(loc="lower right")
plt.grid(linestyle='-.')  
plt.grid(True)
plt.show()


<Figure size 432x288 with 1 Axes>

In [14]:
roc_auc

{0: 0.9743589743589745,
 1: 0.8611111111111112,
 2: 0.9681372549019607,
 'micro': 0.9173111243307555,
 'macro': 0.9387752430031842}

AUC是评估二分类模型性能的重要指标，其值越接近于1，模型的分类能力越好。  
根据ROC曲线和对应的AUC值，我们可以得出以下结论：  

- 对于类别0，ROC AUC为0.97，表示模型在预测类别0时具有非常高的分类能力和区分度。  
- 对于类别1，ROC AUC为0.86，表示模型在预测类别1时的分类能力相对较好，但稍低于类别0。  
- 对于类别2，ROC AUC为0.97，表示模型在预测类别2时具有非常高的分类能力和区分度。  
- 宏平均（macro average）的ROC AUC为0.94，表示模型在各个类别上的平均分类能力很好。  
- 微平均（micro average）的ROC AUC为0.92，表示模型在整体数据集上的分类能力较好。  

综合而言，模型在这些类别上具有非常好的分类能力和区分度。因此，根据给出的测试数据的AUC值，我们可以确认模型在这些类别上的分类能力表现非常出色。  
AUC is an important index to evaluate the performance of the binary classification model, the closer its value is to 1, the better the classification ability of the model.  
According to the ROC curve and the corresponding AUC value, we can draw the following conclusions:  

For category 0, the ROC AUC is 0.97, indicating that the model has very high classification ability and discrimination when predicting category 0.  
For category 1, the ROC AUC is 0.86, indicating that the classification ability of the model is relatively good when predicting category 1, but slightly lower than category 0.  
For category 2, the ROC AUC is 0.97, indicating that the model has very high classification ability and discrimination when predicting category 2.  
The ROC AUC of the macro average (macro average) is 0.94, indicating that the average classification ability of the model on each category is very good.  
The ROC AUC of the micro average is 0.92, indicating that the model has a good classification ability on the overall data set.  
Overall, the model has very good classification and discrimination on these categories. Therefore, based on the AUC values given for the test data, we can confirm that the classification ability of the model on these categories is very good.

In [15]:
# df_changed[df_changed["label"]==0]

**(2) Best correlate with class**与级最相关

In [16]:
df_tmp=df_changed.copy()

# 将标签映射为其他值Map labels to other values
label_map = { 0:1,1:0, 2:0}
df_tmp['label_0'] = df_tmp['label'].map(label_map)

# 将标签映射为其他值Map labels to other values
label_map = { 0:0,1:1, 2:0}
df_tmp['label_1'] = df_tmp['label'].map(label_map)

# 将标签映射为其他值Map labels to other values
label_map = { 0:0,1:0, 2:1}
df_tmp['label_2'] = df_tmp['label'].map(label_map)


df_tmp



0    1    2    3    4    5    6    7    8    9  ...   75   76   77  \
0    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
3    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
4    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
5    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
6    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
7    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
8    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
9    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
10   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
11   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
12   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
13   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
14   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
15   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
16   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
17   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
18   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
19   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
20   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
21   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
22   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
23   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
24   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
25   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
26   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
27   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
28   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
29   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
174  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
175  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
176  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
177  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
178  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
179  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
180  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
181  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
182  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
183  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
184  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  1.0  0.0  0.0   
185  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
186  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
187  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
188  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
189  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
190  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
191  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
192  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
193  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
194

In [17]:
# 计算所有变量与因变量的相关系数Calculate the correlation coefficient of all variables with the dependent variable
correlation = df_tmp.drop(["label","label_1","label_2"],axis=1).corr()['label_0'].abs().sort_values(ascending=False)

# 获取与因变量最相关的三个变量Get the three variables most correlated with the dependent variable
top_3_variables = correlation[1:4].index.tolist()

print("与分类0最相关的三个变量所在的列数分别是The column numbers of the three variables most correlated with category 0 are：", top_3_variables)
print("对应相关系数分别是The corresponding correlation coefficients are：")
correlation[1:4]


与分类0最相关的三个变量所在的列数分别是The column numbers of the three variables most correlated with category 0 are： [39, 40, 41]
对应相关系数分别是The corresponding correlation coefficients are：


39    0.626783
40    0.626783
41    0.626783
Name: label_0, dtype: float64

In [18]:
# 计算所有变量与因变量的相关系数Calculate the correlation coefficient of all variables with the dependent variable
correlation = df_tmp.drop(["label","label_0","label_2"],axis=1).corr()['label_1'].abs().sort_values(ascending=False)

# 获取与因变量最相关的三个变量Get the three variables most correlated with the dependent variable
top_3_variables = correlation[1:4].index.tolist()

print("与分类1最相关的三个变量所在的列数分别是The column numbers of the three variables most relevant to category 1 are：", top_3_variables)
print("对应相关系数分别是The corresponding correlation coefficients are：")
correlation[1:4]

与分类1最相关的三个变量所在的列数分别是The column numbers of the three variables most relevant to category 1 are： [39, 40, 41]
对应相关系数分别是The corresponding correlation coefficients are：


39    0.298807
40    0.298807
41    0.298807
Name: label_1, dtype: float64

In [19]:
# 计算所有变量与因变量的相关系数Calculate the correlation coefficient of all variables with the dependent variable
correlation = df_tmp.drop(["label","label_0","label_1"],axis=1).corr()['label_2'].abs().sort_values(ascending=False)

# 获取与因变量最相关的三个变量Get the three variables most correlated with the dependent variable
top_3_variables = correlation[1:4].index.tolist()

print("与分类2最相关的三个变量所在的列数分别是The column numbers of the three variables that are most relevant to category 2 are：", top_3_variables)
print("对应相关系数分别是The corresponding correlation coefficients are：")
correlation[1:4]


与分类2最相关的三个变量所在的列数分别是The column numbers of the three variables that are most relevant to category 2 are： [37, 68, 66]
对应相关系数分别是The corresponding correlation coefficients are：


37    0.494413
68    0.494413
66    0.494413
Name: label_2, dtype: float64

 #### 1.4.2 **噪音数据**noise data

In [20]:
import warnings 
warnings.filterwarnings('ignore') 
# print(X1.shape,"\n",X1[0,:,:,0])
X1[0,:,:,0].reshape(-1,X1.shape[1]*X1.shape[1])

img = X1[0,:,:,0]
# 应用高斯滤波apply Gaussian filtering
denoised_image = cv2.GaussianBlur(img, (5,5), 0)
# plt.imshow(denoised_image)
# plt.imshow(denoised_image)
# (denoised_image)
X1_changed=pd.DataFrame(denoised_image.reshape(-1,X1.shape[1]*X1.shape[1]))
X1_changed

for i in range(1,X1.shape[0]):
    img = X1[i,:,:,0]
    denoised_image = cv2.GaussianBlur(img, (5,5), 0)
    tmp = pd.DataFrame(denoised_image.reshape(-1,X1.shape[1]*X1.shape[1]))
    #append rows of df2 to end of existing DataFrame
#     X_changed.loc[i]=tmp
    X1_changed = X1_changed.append(tmp)
    X1_changed = X1_changed.reset_index().drop(["index"],axis=1)
#     X_changed=X_changed.reset_index()

    
X1_changed["label"]=y1
df_noise_changed = X1_changed.copy()
df_noise_changed


0         1         2         3         4         5         6  \
0    0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
1    0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
2    0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
3    0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
4    0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
5    0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
6    0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
7    0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
8    0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
9    0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
10   0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
11   0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
12   0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
13   0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
14   0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
15   0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
16   0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
17   0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
18   0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
19   0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
20   0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
21   0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
22   0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
23   0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
24   0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
25   0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
26   0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
27   0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
28   0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
29   0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
..        ...       ...       ...       ...       ...       ...       ...   
174  0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
175  0.015625  0.054688  0.140625  0.171875  0.101562  0.031250  0.031250   
176  0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
177  0.109375  0.148438  0.171875  0.101562  0.023438  0.007812  0.031250   
178  0.250000  0.218750  0.125000  0.039062  0.031250  0.046875  0.031250   
179  0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
180  0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
181  0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
182  0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
183  0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
184  0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
185  0.125000  0.179688  0.218750  0.132812  0.031250  0.007812  0.031250   
186  0.125000  0.179688  0.218750  0.140625  0.062500  0.046875  0.031250   
187  0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
188  0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
189  0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
190  0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
191  0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
192  0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
193  0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
194  0.1

In [21]:

X = df_noise_changed.drop(["label"],axis=1)
y = df_noise_changed.label


# 洗牌和分割训练集和测试集shuffle and split training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.2,random_state=0)

# 创建朴素贝叶斯分类器Create a Naive Bayes Classifier
clf = GaussianNB()

# 拟合模型fit model
clf.fit(X_train, y_train)

# 在测试集上进行预测Make predictions on the test set
y_pred = clf.predict(X_test)

# 计算混淆矩阵Calculate the confusion matrix
cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(cm)

# 计算分类报告Calculate Classification Report
report = classification_report(y_test, y_pred)
print("Classification Report:")
print(report)




Confusion Matrix:
[[12  3  0]
 [ 4  4  1]
 [ 1  5 11]]
Classification Report:
              precision    recall  f1-score   support

         0.0       0.71      0.80      0.75        15
         1.0       0.33      0.44      0.38         9
         2.0       0.92      0.65      0.76        17

    accuracy                           0.66        41
   macro avg       0.65      0.63      0.63        41
weighted avg       0.71      0.66      0.67        41



根据给出的混淆矩阵和分类报告，我们可以得出以下结论：  

- 准确率（accuracy）为0.66，表示模型在测试集上的整体预测准确率为66%。  
- 宏平均（macro average）的精确率为0.65，召回率为0.63，F1-score为0.63。宏平均是对每个类别独立计算指标后取平均值，表示模型在各个类别上的平均表现。  
- 加权平均（weighted average）的精确率为0.71，召回率为0.66，F1-score为0.67。加权平均是按照各类别样本数量的权重进行求平均，表示模型在整体数据集上的加权平均表现。  
- 对于类别0，精确率为0.71，召回率为0.80，F1-score为0.75。模型在预测类别0时的准确率较高，但在召回率上稍有下降。  
- 对于类别1，精确率为0.33，召回率为0.44，F1-score为0.38。模型在预测类别1时的表现较差，准确率、召回率和F1-score都较低。  
- 对于类别2，精确率为0.92，召回率为0.65，F1-score为0.76。模型在预测类别2时的准确率较高，但在召回率上有所下降。  

总体而言，模型在类别0和类别2上表现较好，而在类别1上的预测较不准确。  

According to the given confusion matrix and classification report, we can draw the following conclusions:  

The accuracy rate (accuracy) is 0.66, indicating that the overall prediction accuracy of the model on the test set is 66%.  
The precision of the macro average is 0.65, the recall is 0.63, and the F1-score is 0.63. Macro averaging is to calculate the average value for each category independently, indicating the average performance of the model on each category.  
The weighted average has a precision of 0.71, a recall of 0.66, and an F1-score of 0.67. The weighted average is averaged according to the weight of the number of samples in each category, indicating the weighted average performance of the model on the overall data set.  
For category 0, the precision is 0.71, the recall is 0.80, and the F1-score is 0.75. The model has higher precision when predicting class 0, but a slight drop in recall.  
For category 1, the precision is 0.33, the recall is 0.44, and the F1-score is 0.38. The model performs poorly in predicting class 1, with low precision, recall and F1-score.  
For category 2, the precision is 0.92, the recall is 0.65, and the F1-score is 0.76. The model has higher precision when predicting class 2, but a drop in recall.  
Overall, the model performs better on classes 0 and 2, while predicting less accurately on class 1.

In [22]:
# 在测试集上进行预测Make predictions on the test set
y_pred = clf.predict_proba(X_test)
y_pred

classes=[0,1,2]

y_test_all = label_binarize(y_test, classes=[0,1,2])

# 在测试集上进行预测Make predictions on the test set
y_pred = clf.predict_proba(X_test)
y_pred

y_score_all=y_pred
fpr = dict()
tpr = dict()
roc_auc = dict()
for i in range(len(classes)):
    fpr[i], tpr[i], thresholds = roc_curve(y_test_all[:, i],y_score_all[:, i])
    roc_auc[i] = auc(fpr[i], tpr[i])

roc_auc

# micro-average ROC curve（方法一）
fpr["micro"], tpr["micro"], thresholds = roc_curve(y_test_all.ravel(),y_pred.ravel())
roc_auc["micro"] = auc(fpr["micro"], tpr["micro"])
roc_auc



# macro-average ROC curve 方法二）

all_fpr = np.unique(np.concatenate([fpr[i] for i in range(len(classes))]))

mean_tpr = np.zeros_like(all_fpr)
for i in range(len(classes)):
    mean_tpr += interp(all_fpr, fpr[i], tpr[i])
# 求平均计算ROC包围的面积AUC(Calculate the average area AUC surrounded by ROC
mean_tpr /= len(classes)
fpr["macro"] = all_fpr
tpr["macro"] = mean_tpr
roc_auc["macro"] = auc(fpr["macro"], tpr["macro"])

#画图部分drawing part
plt.figure()
plt.plot(fpr["micro"], tpr["micro"],'k-',color='y',
         label='ROC curve micro-average(AUC = {0:0.4f})'
               ''.format(roc_auc["micro"]),
          linestyle='-.', linewidth=3)

plt.plot(fpr["macro"], tpr["macro"],'k-',color='k',
         label='ROC curve macro-average(AUC = {0:0.4f})'
               ''.format(roc_auc["macro"]),
          linestyle='-.', linewidth=3)
plt.plot([0,1],[0,1],'r--')
plt.xlim([-0.1,1.1])
plt.ylim([-0.1,1.1])
plt.ylabel('True Positive Rate')
plt.xlabel('False Positive Rate')
plt.legend(loc="lower right")
plt.grid(linestyle='-.')  
plt.grid(True)
plt.show()


<Figure size 432x288 with 1 Axes>

In [23]:
roc_auc


{0: 0.9333333333333333,
 1: 0.7673611111111112,
 2: 0.8627450980392156,
 'micro': 0.8667459845330161,
 'macro': 0.8721658357075024}

根据给出的ROC AUC（Receiver Operating Characteristic Area Under Curve）值，我们可以得出以下结论：  

- 对于类别0，ROC AUC为0.93，表示模型在预测类别0时具有较高的分类能力和区分度。  
- 对于类别1，ROC AUC为0.77，表示模型在预测类别1时的分类能力稍低于类别0，但仍具有一定的区分度。  
- 对于类别2，ROC AUC为0.86，表示模型在预测类别2时具有较好的分类能力和区分度。  
- 宏平均（macro average）的ROC AUC为0.87，表示模型在各个类别上的平均分类能力较好。  
- 微平均（micro average）的ROC AUC为0.87，表示模型在整体数据集上的分类能力较好。  

综合而言，模型在整体数据集上具有较好的分类能力，并且对于类别0、类别1和类别2分别表现出不同程度的区分度。  
ROC AUC是评估二分类模型性能的重要指标，其值越接近于1，模型的分类能力越好。  

因此，根据给出的ROC曲线和AUC值，我们可以得出结论：模型在这些类别上具有良好的分类能力。  

According to the given ROC AUC (Receiver Operating Characteristic Area Under Curve) value, we can draw the following conclusions:  

For category 0, the ROC AUC is 0.93, indicating that the model has high classification ability and discrimination when predicting category 0.  
For category 1, the ROC AUC is 0.77, indicating that the classification ability of the model in predicting category 1 is slightly lower than that of category 0, but still has a certain degree of discrimination.  
For category 2, the ROC AUC is 0.86, indicating that the model has good classification ability and discrimination when predicting category 2.  
The ROC AUC of the macro average (macro average) is 0.87, indicating that the average classification ability of the model in each category is better.  
The ROC AUC of the micro average is 0.87, indicating that the model has a good classification ability on the overall data set.  
In general, the model has a good classification ability on the overall data set, and shows different degrees of discrimination for category 0, category 1 and category 2 respectively.  
ROC AUC is an important indicator to evaluate the performance of the binary classification model, the closer its value is to 1, the better the classification ability of the model.  

Therefore, according to the given ROC curve and AUC value, we can conclude that the model has good classification ability on these categories.

In [24]:
df_tmp=df_noise_changed.copy()

# 将标签映射为其他值Map labels to other values
label_map = { 0:1,1:0, 2:0}
df_tmp['label_0'] = df_tmp['label'].map(label_map)

# 将标签映射为其他值Map labels to other values
label_map = { 0:0,1:1, 2:0}
df_tmp['label_1'] = df_tmp['label'].map(label_map)

# 将标签映射为其他值Map labels to other values
label_map = { 0:0,1:0, 2:1}
df_tmp['label_2'] = df_tmp['label'].map(label_map)


df_tmp



0         1         2         3         4         5         6  \
0    0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
1    0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
2    0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
3    0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
4    0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
5    0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
6    0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
7    0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
8    0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
9    0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
10   0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
11   0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
12   0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
13   0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
14   0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
15   0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
16   0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
17   0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
18   0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
19   0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
20   0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
21   0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
22   0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
23   0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
24   0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
25   0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
26   0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
27   0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
28   0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
29   0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
..        ...       ...       ...       ...       ...       ...       ...   
174  0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
175  0.015625  0.054688  0.140625  0.171875  0.101562  0.031250  0.031250   
176  0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
177  0.109375  0.148438  0.171875  0.101562  0.023438  0.007812  0.031250   
178  0.250000  0.218750  0.125000  0.039062  0.031250  0.046875  0.031250   
179  0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
180  0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
181  0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
182  0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
183  0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
184  0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
185  0.125000  0.179688  0.218750  0.132812  0.031250  0.007812  0.031250   
186  0.125000  0.179688  0.218750  0.140625  0.062500  0.046875  0.031250   
187  0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
188  0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
189  0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
190  0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
191  0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
192  0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
193  0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
194  0.1

**(2) Best correlate with class**与级最相关

In [25]:
df_tmp=df_noise_changed.copy()

# 将标签映射为其他值Map labels to other values
label_map = { 0:1,1:0, 2:0}
df_tmp['label_0'] = df_tmp['label'].map(label_map)

# 将标签映射为其他值Map labels to other values
label_map = { 0:0,1:1, 2:0}
df_tmp['label_1'] = df_tmp['label'].map(label_map)

# 将标签映射为其他值Map labels to other values
label_map = { 0:0,1:0, 2:1}
df_tmp['label_2'] = df_tmp['label'].map(label_map)


df_tmp



0         1         2         3         4         5         6  \
0    0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
1    0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
2    0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
3    0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
4    0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
5    0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
6    0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
7    0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
8    0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
9    0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
10   0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
11   0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
12   0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
13   0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
14   0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
15   0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
16   0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
17   0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
18   0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
19   0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
20   0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
21   0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
22   0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
23   0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
24   0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
25   0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
26   0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
27   0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
28   0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
29   0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
..        ...       ...       ...       ...       ...       ...       ...   
174  0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
175  0.015625  0.054688  0.140625  0.171875  0.101562  0.031250  0.031250   
176  0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
177  0.109375  0.148438  0.171875  0.101562  0.023438  0.007812  0.031250   
178  0.250000  0.218750  0.125000  0.039062  0.031250  0.046875  0.031250   
179  0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
180  0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
181  0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
182  0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
183  0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
184  0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
185  0.125000  0.179688  0.218750  0.132812  0.031250  0.007812  0.031250   
186  0.125000  0.179688  0.218750  0.140625  0.062500  0.046875  0.031250   
187  0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
188  0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
189  0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
190  0.062500  0.054688  0.031250  0.015625  0.031250  0.046875  0.031250   
191  0.015625  0.031250  0.046875  0.031250  0.007812  0.007812  0.031250   
192  0.015625  0.031250  0.046875  0.031250  0.015625  0.031250  0.046875   
193  0.062500  0.054688  0.031250  0.007812  0.000000  0.007812  0.031250   
194  0.1

In [26]:
# 计算所有变量与因变量的相关系数Calculate the correlation coefficient of all variables with the dependent variable
correlation = df_tmp.drop(["label","label_1","label_2"],axis=1).corr()['label_0'].abs().sort_values(ascending=False)

# 获取与因变量最相关的三个变量Get the three variables most correlated with the dependent variable
top_6_variables = correlation[1:7].index.tolist()

print("与分类0最相关的六个变量所在的列数分别是The column numbers of the six variables that are most relevant to category 0 are：", top_6_variables)
print("对应相关系数分别是The corresponding correlation coefficients are：")
correlation[1:7]


与分类0最相关的六个变量所在的列数分别是The column numbers of the six variables that are most relevant to category 0 are： [40, 58, 31, 41, 39, 67]
对应相关系数分别是The corresponding correlation coefficients are：


40    0.676434
58    0.653886
31    0.642888
41    0.636677
39    0.620580
67    0.602882
Name: label_0, dtype: float64

In [27]:
# 计算所有变量与因变量的相关系数Calculate the correlation coefficient of all variables with the dependent variable
correlation = df_tmp.drop(["label","label_0","label_2"],axis=1).corr()['label_1'].abs().sort_values(ascending=False)

# 获取与因变量最相关的三个变量Get the three variables most correlated with the dependent variable
top_6_variables = correlation[1:7].index.tolist()

print("与分类1最相关的六个变量所在的列数分别是The number of columns in which the six variables most correlated with category 1 are：", top_6_variables)
print("对应相关系数分别是The corresponding correlation coefficients are：")
correlation[1:7]


与分类1最相关的六个变量所在的列数分别是The number of columns in which the six variables most correlated with category 1 are： [33, 43, 44, 37, 29, 42]
对应相关系数分别是The corresponding correlation coefficients are：


33    0.312833
43    0.298609
44    0.279782
37    0.274949
29    0.267561
42    0.253570
Name: label_1, dtype: float64

In [28]:
# 计算所有变量与因变量的相关系数Calculate the correlation coefficient of all variables with the dependent variable
correlation = df_tmp.drop(["label","label_1","label_0"],axis=1).corr()['label_2'].abs().sort_values(ascending=False)

# 获取与因变量最相关的三个变量Get the three variables most correlated with the dependent variable
top_6_variables = correlation[1:7].index.tolist()

print("与分类2最相关的六个变量所在的列数分别是The number of columns of the six variables most relevant to category 2 are：", top_6_variables)
print("对应相关系数分别是The corresponding correlation coefficients are：")
correlation[1:7]



与分类2最相关的六个变量所在的列数分别是The number of columns of the six variables most relevant to category 2 are： [67, 76, 40, 68, 66, 77]
对应相关系数分别是The corresponding correlation coefficients are：


67    0.651639
76    0.609752
40    0.608104
68    0.598697
66    0.589130
77    0.585593
Name: label_2, dtype: float64

In [29]:
reset -f

In [30]:
# correlation

## Part 2.  Clustering

### 2.1 Kmeans聚类

In [31]:
# df_changed.drop(["label"],axis=1)

In [32]:


import numpy as np # linear algebra线性代数
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)数据处理、CSV 文件 I/O（例如 pd.read_csv）
import numpy as np
import matplotlib.pyplot as plt
import os
from sklearn.cluster import KMeans
import numpy as np
from sklearn.cluster import KMeans
from sklearn.metrics import adjusted_rand_score
import numpy as np


X = np.load('/home/mw/input/smiley2472/smiley_X.npy')
y = np.load('/home/mw/input/smiley2472/smiley_Y.npy')

import warnings 
X_changed=pd.DataFrame(X[0,:,:,0].reshape(-1,X.shape[1]*X.shape[1]))

for i in range(1,X.shape[0]):
    tmp = pd.DataFrame(X[i,:,:,0].reshape(-1,X.shape[1]*X.shape[1]))
    X_changed = X_changed.append(tmp)
    X_changed = X_changed.reset_index().drop(["index"],axis=1)

X_changed["label"]=y
df_changed = X_changed.copy()
df_changed.head()

# print(df_changed.label)
# df_changed.label.tolist()[:]

# 创建 KMeans 模型并指定聚类簇的数量为 k/Create a KMeans model and specify the number of clusters as k
k = 3
seed = 123
kmeans = KMeans(n_clusters=k, random_state=seed)

df_changed.drop(["label"],axis=1)
# 在数据上执行 K-means 聚类Perform K-means clustering on the data
kmeans.fit(df_changed.drop(["label"],axis=1))

# 获取每个样本所属的簇标签Get the cluster label to which each sample belongs
labels = kmeans.labels_

# 获取每个簇的中心点坐标Get the center point coordinates of each cluster
centers = kmeans.cluster_centers_

# print("样本聚类标签：", labels)

df_changed["cluster"]=labels
df_changed



0    1    2    3    4    5    6    7    8    9  ...   73   74   75  \
0    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
3    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
4    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
5    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
6    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
7    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
8    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
9    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
10   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
11   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
12   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
13   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
14   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
15   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
16   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
17   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
18   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
19   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
20   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
21   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
22   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
23   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
24   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
25   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
26   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
27   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
28   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
29   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
174  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
175  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
176  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
177  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
178  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
179  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  1.0  0.0  0.0   
180  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
181  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  1.0  0.0   
182  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  1.0  0.0   
183  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
184  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  1.0   
185  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
186  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
187  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
188  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
189  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
190  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
191  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
192  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
193  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
194

In [33]:


# 将标签映射为其他值Map labels to other values
label_map = { 0:1,1:2, 2:0} 

df_changed['cluster'] = df_changed['cluster'].map(label_map)
df_changed


from sklearn.metrics import classification_report, roc_curve, auc

# 计算分类报告Calculate Classification Report
report = classification_report(df_changed.label, df_changed.cluster)
print("Classification Report:")
print(report)



Classification Report:
              precision    recall  f1-score   support

         0.0       1.00      0.50      0.67        72
         1.0       0.43      0.60      0.50        60
         2.0       0.71      0.83      0.77        72

    accuracy                           0.65       204
   macro avg       0.71      0.64      0.65       204
weighted avg       0.73      0.65      0.65       204



### 2.2 Kmeans聚类可视化 /Kmeans clustering visualization

In [34]:


import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
import numpy as np

# 创建 PCA 模型并设置主成分数目Create a PCA model and set the number of principal components
n_components = 2
pca = PCA(n_components=n_components)

X = df_changed.iloc[:,:81]
# 在数据上执行 PCAPerform PCA on the data
pca.fit(X)

# 将数据转换到降维后的空间Transform the data into the reduced dimension space
X_pca = pca.transform(X)

# 创建 KMeans 模型并设置聚类数目Create a KMeans model and set the number of clusters
k = 3
kmeans = KMeans(n_clusters=k, random_state=123)

# 在降维后的数据上执行 K-means 聚类Perform K-means clustering on the reduced data
kmeans.fit(X_pca)

# 获取每个样本所属的簇标签Get the cluster label to which each sample belongs
labels = kmeans.labels_

# 获取聚类中心点的坐标Get the coordinates of the cluster center point
centroids = kmeans.cluster_centers_

# 绘制散点图Draw a scatterplot
plt.scatter(X_pca[:, 0], X_pca[:, 1], c=labels)
# 绘制聚类中心点Draw cluster centroids
plt.scatter(centroids[:, 0], centroids[:, 1], marker='x', color='red')
plt.xlabel('PC1')
plt.ylabel('PC2')
plt.title('K-means Clustering Visualize')
plt.show()

<Figure size 432x288 with 1 Axes>

### 2.3 层次聚类（hierarchical clustering ）

In [35]:
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
from sklearn.cluster import AgglomerativeClustering
import numpy as np

X = df_changed.iloc[:,:81]
# 执行层次聚类perform hierarchical clustering
n_clusters = 3
clustering = AgglomerativeClustering(n_clusters=n_clusters)


# 执行层次聚类perform hierarchical clustering
labels = clustering.fit_predict(X)
labels


X["cluster"]=labels
X["label"]=df_changed.label

# 标签映射label mapping
label_map = { 0:2,1:1, 2:0} 

X['cluster'] = X['cluster'].map(label_map)
X


from sklearn.metrics import classification_report, roc_curve, auc

# 计算分类报告Calculate Classification Report
report = classification_report(X.label, X.cluster)
print("Classification Report:")
print(report)



Classification Report:
              precision    recall  f1-score   support

         0.0       1.00      0.50      0.67        72
         1.0       0.43      0.60      0.50        60
         2.0       0.71      0.83      0.77        72

    accuracy                           0.65       204
   macro avg       0.71      0.64      0.65       204
weighted avg       0.73      0.65      0.65       204



这里可以看到，层次聚类和Kmeans结果的准确率是一致的。  
It can be seen here that the accuracy of hierarchical clustering and Kmeans results is consistent

### 2.4 Elbow Method确定最优聚类个数  (Elbow Method to determine the optimal number of clusters)  

确定最优聚类个数的一种常用方法是使用K均值聚类（K-means clustering）的“手肘法”（Elbow Method）。  

手肘法基于聚类结果的变异程度来选择最优的聚类个数。当聚类个数增加时，每个样本点与其所属聚类中心的距离会减小，导致总体的误差减小。然而，随着聚类个数继续增加，每个新增的聚类中心所带来的误差减少逐渐减缓，最终达到一个拐点。这个拐点称为“手肘”，它反映了在聚类个数增加过程中，聚类效果的显著改善程度逐渐降低。  
A common method to determine the optimal number of clusters is to use the "Elbow Method" of K-means clustering.  

The elbow method selects the optimal number of clusters based on the degree of variation of the clustering results. When the number of clusters increases, the distance between each sample point and its cluster center will decrease, resulting in a decrease in the overall error. However, as the number of clusters continues to increase, the error reduction brought by each new cluster center gradually slows down, and finally reaches an inflection point. This inflection point is called "elbow", which reflects that the degree of significant improvement in clustering effect gradually decreases as the number of clusters increases.  


In [36]:
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans
from sklearn.datasets import make_blobs
from sklearn.preprocessing import StandardScaler

X = df_changed.iloc[:,:81]
# 标准化数据normalized data
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# 设置不同的聚类个数进行试验Experiment with different numbers of clusters
k_values = range(1, 9)
sse = []

for k in k_values:
    kmeans = KMeans(n_clusters=k,random_state=123)
    kmeans.fit(X_scaled)
    sse.append(kmeans.inertia_)

# 绘制手肘法图形Drawing the Elbow Method Graphics
plt.plot(k_values, sse, 'bx-')
plt.xlabel('Number of Clusters')
plt.ylabel('SSE (Sum of Squared Errors)')
plt.title('Elbow Method with Standardized Data')
plt.show()


<Figure size 432x288 with 1 Axes>

不难看出，拐点在$k=4$附近，这说明该数据的最优聚类个数为4个  
The optimal number of clusters for this data is 4

In [37]:
# 创建 KMeans 模型并指定聚类簇的数量为 k/Create a KMeans model and specify the number of clusters as k
k = 4
seed = 123
kmeans = KMeans(n_clusters=k, random_state=seed)

X = df_changed.iloc[:,:81]
# 在数据上执行 K-means 聚类Perform K-means clustering on the data
kmeans.fit(X)

# 获取每个样本所属的簇标签Get the cluster label to which each sample belongs
labels = kmeans.labels_

# 获取每个簇的中心点坐标Get the center point coordinates of each cluster
centers = kmeans.cluster_centers_

# print("样本聚类标签：", labels)

X["cluster"]=labels

X["label"]=df_changed.label

# 标签映射label mapping
label_map = { 0:1,1:0, 2:2,3:3} 

X['cluster'] = X['cluster'].map(label_map)
X



0    1    2    3    4    5    6    7    8    9  ...   73   74   75  \
0    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
3    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
4    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
5    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
6    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
7    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
8    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
9    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
10   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
11   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
12   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
13   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
14   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
15   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
16   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
17   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
18   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
19   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
20   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
21   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
22   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
23   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
24   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
25   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
26   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
27   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
28   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
29   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
174  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
175  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
176  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
177  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
178  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
179  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  1.0  0.0  0.0   
180  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
181  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  1.0  0.0   
182  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  1.0  0.0   
183  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
184  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  1.0   
185  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
186  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
187  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
188  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
189  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
190  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
191  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
192  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
193  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
194

In [38]:
# 计算分类报告Calculate Classification Report
report = classification_report(X.label,X.cluster)
print("Classification Report:")
print(report)



Classification Report:
              precision    recall  f1-score   support

         0.0       1.00      0.50      0.67        72
         1.0       0.40      0.40      0.40        60
         2.0       1.00      0.33      0.50        72
         3.0       0.00      0.00      0.00         0

    accuracy                           0.41       204
   macro avg       0.60      0.31      0.39       204
weighted avg       0.82      0.41      0.53       204



结果分析：  

- **准确率**（precision）：从整体来看，模型的准确率在0.40到1.00之间，其中类别0和类别2的准确率较高。  

- **召回率**（recall）：在这里，召回率的范围从0.00到0.50不等，说明模型对于某些类别的判断能力较弱。  

- **F1得分**（F1-score）：整体来看，F1得分在0.00到0.67之间，说明模型的性能存在一定的差异。  


总体分析：从宏平均和加权平均的角度来看，模型的平均准确率、平均召回率和平均F1得分大致在0.31到0.60之间，整体表现一般。  

综上所述，可以看出这里如果聚类成4类，**会导致模型预测出多余的一个类别，导致算法准确率进一步降低**。  
Result analysis:  

Accuracy (precision): On the whole, the accuracy of the model is between 0.40 and 1.00, and the accuracy of category 0 and category 2 is higher.  

Recall (recall): Here, the recall rate ranges from 0.00 to 0.50, indicating that the model's ability to judge certain categories is weak.  

F1 score (F1-score): Overall, the F1 score is between 0.00 and 0.67, indicating that there are certain differences in the performance of the model.  

Overall analysis: From the perspective of macro average and weighted average, the average accuracy, average recall, and average F1 score of the model are roughly between 0.31 and 0.60, and the overall performance is average.  

To sum up, it can be seen that if clustering into 4 categories here, the model will predict an extra category, which will further reduce the accuracy of the algorithm.

In [39]:
reset -f

## Part 3.  Decision trees决策树

### 3.1 Decision tree

In [40]:


import numpy as np # linear algebra线性代数
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)数据处理、CSV 文件 I/O（例如 pd.read_csv）
import numpy as np
import matplotlib.pyplot as plt
import os
from sklearn.cluster import KMeans
import numpy as np
from sklearn.cluster import KMeans
from sklearn.metrics import adjusted_rand_score
import numpy as np
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report, roc_curve, auc
import numpy as np
import matplotlib.pyplot as plt
from itertools import combinations
# 引入必要的库
import numpy as np
import matplotlib.pyplot as plt
from itertools import cycle
from sklearn import svm, datasets
from sklearn.metrics import roc_curve, auc
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import label_binarize
from sklearn.multiclass import OneVsRestClassifier
from scipy import interp
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report, confusion_matrix, roc_curve, auc
import numpy as np
import matplotlib.pyplot as plt

X = np.load('/home/mw/input/smiley2472/smiley_X.npy')
y = np.load('/home/mw/input/smiley2472/smiley_Y.npy')

import warnings 
X_changed=pd.DataFrame(X[0,:,:,0].reshape(-1,X.shape[1]*X.shape[1]))

for i in range(1,X.shape[0]):
    tmp = pd.DataFrame(X[i,:,:,0].reshape(-1,X.shape[1]*X.shape[1]))
    X_changed = X_changed.append(tmp)
    X_changed = X_changed.reset_index().drop(["index"],axis=1)

X_changed["label"]=y
df_changed = X_changed.copy()
df_changed.head()

from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report


X = df_changed.drop(["label"],axis=1)
y = df_changed.label

# 将数据集分为训练集和测试集Divide the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 创建决策树分类器对象Create a decision tree classifier object
clf = DecisionTreeClassifier()

# 在训练集上拟合决策树模型Fit a decision tree model on the training set
clf.fit(X_train, y_train)

# 在测试集上进行预测Make predictions on the test set
y_pred = clf.predict(X_train)

print("训练集分类报告Training Set Classification Report:\n")
# 输出分类报告output classification report
print(classification_report(y_train, y_pred))



训练集分类报告Training Set Classification Report:

              precision    recall  f1-score   support

         0.0       1.00      1.00      1.00        57
         1.0       1.00      1.00      1.00        51
         2.0       1.00      1.00      1.00        55

    accuracy                           1.00       163
   macro avg       1.00      1.00      1.00       163
weighted avg       1.00      1.00      1.00       163



这里发现，**对于训练集，模型的准确率为100%**。Here it is found that **For the training set, the accuracy of the model is 100%**.

In [41]:
# 在测试集上进行预测Make predictions on the test set
y_pred = clf.predict(X_test)

print("测试集分类报告Test Set Classification Report：\n")
# 输出分类报告output classification report
print(classification_report(y_test, y_pred,digits=4))


# 在测试集上进行预测Make predictions on the test set
y_pred = clf.predict_proba(X_test)
y_pred

classes=[0,1,2]
y_test_all = label_binarize(y_test, classes=[0,1,2])

# 在测试集上进行预测Make predictions on the test set
y_pred = clf.predict_proba(X_test)
y_pred

y_score_all=y_pred
fpr = dict()
tpr = dict()
roc_auc = dict()
for i in range(len(classes)):
    fpr[i], tpr[i], thresholds = roc_curve(y_test_all[:, i],y_score_all[:, i])
    roc_auc[i] = auc(fpr[i], tpr[i])


# micro-average ROC curve（方法一）
fpr["micro"], tpr["micro"], thresholds = roc_curve(y_test_all.ravel(),y_pred.ravel())
roc_auc["micro"] = auc(fpr["micro"], tpr["micro"])


# macro-average ROC curve 方法二）
all_fpr = np.unique(np.concatenate([fpr[i] for i in range(len(classes))]))

mean_tpr = np.zeros_like(all_fpr)
for i in range(len(classes)):
    mean_tpr += interp(all_fpr, fpr[i], tpr[i])
# 求平均计算ROC包围的面积AUC/ Calculate the average area AUC surrounded by ROC
mean_tpr /= len(classes)
fpr["macro"] = all_fpr
tpr["macro"] = mean_tpr
roc_auc["macro"] = auc(fpr["macro"], tpr["macro"])

#画图部分drawing part
plt.figure()
plt.plot(fpr["micro"], tpr["micro"],'k-',color='y',
         label='ROC curve micro-average(AUC = {0:0.4f})'
               ''.format(roc_auc["micro"]),
          linestyle='-.', linewidth=3)

plt.plot(fpr["macro"], tpr["macro"],'k-',color='k',
         label='ROC curve macro-average(AUC = {0:0.4f})'
               ''.format(roc_auc["macro"]),
          linestyle='-.', linewidth=3)
plt.plot([0,1],[0,1],'r--')
plt.xlim([-0.1,1.1])
plt.ylim([-0.1,1.1])
plt.ylabel('True Positive Rate')
plt.xlabel('False Positive Rate')
plt.legend(loc="lower right")
plt.grid(linestyle='-.')  
plt.grid(True)
plt.show()


测试集分类报告Test Set Classification Report：

              precision    recall  f1-score   support

         0.0     1.0000    1.0000    1.0000        15
         1.0     0.9000    1.0000    0.9474         9
         2.0     1.0000    0.9412    0.9697        17

    accuracy                         0.9756        41
   macro avg     0.9667    0.9804    0.9724        41
weighted avg     0.9780    0.9756    0.9759        41



<Figure size 432x288 with 1 Axes>

测试集准确率为：**97.56%, 表明模型性能非常好**，但是和训练集的100%准确率相比有点逊色，这说明模型有点**过拟合**。  
The accuracy of the test set is 97.56%, indicating that the performance of the model is very good, but it is a bit inferior to the 100% accuracy of the training set, which shows that the model is a bit overfitting.

### 3.2 Decision tree + cv

In [42]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.datasets import make_blobs
from sklearn.model_selection import GridSearchCV

X = df_changed.drop(["label"],axis=1)
y = df_changed.label

# 将数据集分为训练集和测试集Divide the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=123)

# 创建决策树分类器对象Create a decision tree classifier object
tree = DecisionTreeClassifier(random_state=123)

# 定义要进行交叉验证的参数候选值Define parameter candidates to be cross-validated
param_grid = {
    'max_depth': [2,3,5,7,9,10],
    'min_samples_split': [2,3,4,5,6,7,8,9,10],
    'min_samples_leaf': [2, 3,4,5],
    'criterion': ['gini', 'entropy']
}

# 定义交叉验证对象Define the cross-validation object
grid_search = GridSearchCV(tree, param_grid, cv=10)

# 执行交叉验证和参数调优Perform cross-validation and parameter tuning
grid_search.fit(X_train, y_train)

# 输出最佳模型的参数Output the parameters of the best model
print("cv Best Parameters:", grid_search.best_params_)

# 输出最佳模型的得分output the score of the best model
print("cv Best Score:", grid_search.best_score_)


cv Best Parameters: {'criterion': 'entropy', 'max_depth': 7, 'min_samples_leaf': 2, 'min_samples_split': 2}
cv Best Score: 0.9754601226993865


In [43]:


y_pred = grid_search.predict(X_test)

print("测试集分类报告Test Set Classification Report：\n")
# 输出分类报告output classification report
print(classification_report(y_test, y_pred,digits=4))




测试集分类报告Test Set Classification Report：

              precision    recall  f1-score   support

         0.0     1.0000    1.0000    1.0000        11
         1.0     0.9231    1.0000    0.9600        12
         2.0     1.0000    0.9444    0.9714        18

    accuracy                         0.9756        41
   macro avg     0.9744    0.9815    0.9771        41
weighted avg     0.9775    0.9756    0.9757        41



根据交叉验证集和测试集的分类报告和最佳模型的参数，我们可以得出以下结论：  

- 精确度（precision）和召回率（recall）：各个类别的精确度和召回率指标都较高，表明模型在测试集上具有较好的分类性能。特别是类别0和类别2的精确度和召回率均为1.0，说明模型在这两个类别上准确率和召回率都达到了最高水平。  

- F1 分数（f1-score）：各个类别的 F1 分数也都较高，说明模型在同时考虑准确率和召回率时取得了良好的平衡。  

- 准确率（accuracy）：整体测试集的准确率为0.9756，也很高。这意味着模型能够正确分类大多数的样本。  

- 最佳参数：经过交叉验证和参数调优，最佳模型的参数为  
{'criterion': 'entropy', 'max_depth': None, 'min_samples_leaf': 2, 'min_samples_split': 2}。这些参数是通过在训练集上尝试不同的值来获得的。  

综上所述，基于给定的数据集，经过参数调优后的决策树分类器在**交叉验证集和测试集**上的表现相差非常小(0.9765 v.s 0.9756)，表现都相当优秀，能够对样本进行准确的分类。  

Based on the classification reports of the cross-validation and test sets and the parameters of the best model, we can draw the following conclusions:  

Precision and recall: The precision and recall indicators for each category are high, indicating that the model has good classification performance on the test set. In particular, both category 0 and category 2 have a precision and recall of 1.0, indicating that the model has reached the highest level of precision and recall for both categories.  

F1 score (f1-score): The F1 score of each category is also high, indicating that the model has achieved a good balance when considering both precision and recall.  

Accuracy: The accuracy of the overall test set is 0.9756, which is also very high. This means that the model was able to correctly classify most of the samples.  

Optimal parameters: After cross-validation and parameter tuning, the parameters of the optimal model are  
{'criterion': 'entropy', 'max_depth': None, 'min_samples_leaf': 2, 'min_samples_split': 2}. These parameters are obtained by trying different values on the training set.  

To sum up, based on the given data set, the performance difference between the cross-validation set and the test set of the decision tree classifier after parameter tuning is very small (0.9765 v.s 0.9756), and the performance is quite excellent. accurate classification.

In [44]:

# 在测试集上进行预测Make predictions on the test set
y_pred = grid_search.predict_proba(X_test)
y_pred


classes=[0,1,2]
y_test_all = label_binarize(y_test, classes=[0,1,2])

y_score_all=y_pred
fpr = dict()
tpr = dict()
roc_auc = dict()
for i in range(len(classes)):
    fpr[i], tpr[i], thresholds = roc_curve(y_test_all[:, i],y_score_all[:, i])
    roc_auc[i] = auc(fpr[i], tpr[i])


# micro-average ROC curve（方法一）
fpr["micro"], tpr["micro"], thresholds = roc_curve(y_test_all.ravel(),y_pred.ravel())
roc_auc["micro"] = auc(fpr["micro"], tpr["micro"])


# macro-average ROC curve 方法二）
all_fpr = np.unique(np.concatenate([fpr[i] for i in range(len(classes))]))

mean_tpr = np.zeros_like(all_fpr)
for i in range(len(classes)):
    mean_tpr += interp(all_fpr, fpr[i], tpr[i])
# 求平均计算ROC包围的面积AUC/ Calculate the average area AUC surrounded by ROC
mean_tpr /= len(classes)
fpr["macro"] = all_fpr
tpr["macro"] = mean_tpr
roc_auc["macro"] = auc(fpr["macro"], tpr["macro"])

#画图部分drawing part
plt.figure()
plt.plot(fpr["micro"], tpr["micro"],'k-',color='y',
         label='ROC curve micro-average(AUC = {0:0.4f})'
               ''.format(roc_auc["micro"]),
          linestyle='-.', linewidth=3)

plt.plot(fpr["macro"], tpr["macro"],'k-',color='k',
         label='ROC curve macro-average(AUC = {0:0.4f})'
               ''.format(roc_auc["macro"]),
          linestyle='-.', linewidth=3)
plt.plot([0,1],[0,1],'r--')
plt.xlim([-0.1,1.1])
plt.ylim([-0.1,1.1])
plt.ylabel('True Positive Rate')
plt.xlabel('False Positive Rate')
plt.legend(loc="lower right")
plt.grid(linestyle='-.')  
plt.grid(True)
plt.show()




<Figure size 432x288 with 1 Axes>

In [45]:
reset -f

### 3.3 Decision tree(sample 40%)

In [46]:


import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import numpy as np
import matplotlib.pyplot as plt
import os
from sklearn.metrics import adjusted_rand_score
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report, roc_curve, auc
import numpy as np
import matplotlib.pyplot as plt
from itertools import combinations
# 引入必要的库
import numpy as np
import matplotlib.pyplot as plt
from itertools import cycle
from sklearn import svm, datasets
from sklearn.metrics import roc_curve, auc
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import label_binarize
from sklearn.multiclass import OneVsRestClassifier
from scipy import interp
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report, confusion_matrix, roc_curve, auc
import numpy as np
import matplotlib.pyplot as plt

X = np.load('/home/mw/input/smiley2472/smiley_X.npy')
y = np.load('/home/mw/input/smiley2472/smiley_Y.npy')

import warnings 
X_changed=pd.DataFrame(X[0,:,:,0].reshape(-1,X.shape[1]*X.shape[1]))

for i in range(1,X.shape[0]):
    tmp = pd.DataFrame(X[i,:,:,0].reshape(-1,X.shape[1]*X.shape[1]))
    X_changed = X_changed.append(tmp)
    X_changed = X_changed.reset_index().drop(["index"],axis=1)

X_changed["label"]=y
df_changed = X_changed.copy()
df_changed.head()

from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report


X = df_changed.drop(["label"],axis=1)
y = df_changed.label

# 将数据集分为训练集和测试集Divide the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.6, random_state=42)

# 创建决策树分类器对象Create a decision tree classifier object
clf = DecisionTreeClassifier()

# 在训练集上拟合决策树模型Fit a decision tree model on the training set
clf.fit(X_train, y_train)

# 在测试集上进行预测Make predictions on the test set
y_pred = clf.predict(X_train)

print("训练集分类报告Training Set Classification Report:\n")
# 输出分类报告
print(classification_report(y_train, y_pred))


# 在测试集上进行预测Make predictions on the test set
y_pred = clf.predict(X_test)

print("测试集分类报告Training Set Classification Report：\n")
# 输出分类报告
print(classification_report(y_test, y_pred,digits=4))




训练集分类报告Training Set Classification Report:

              precision    recall  f1-score   support

         0.0       1.00      1.00      1.00        32
         1.0       1.00      1.00      1.00        21
         2.0       1.00      1.00      1.00        28

    accuracy                           1.00        81
   macro avg       1.00      1.00      1.00        81
weighted avg       1.00      1.00      1.00        81

测试集分类报告Training Set Classification Report：

              precision    recall  f1-score   support

         0.0     0.7755    0.9500    0.8539        40
         1.0     0.9333    0.7179    0.8116        39
         2.0     0.9773    0.9773    0.9773        44

    accuracy                         0.8862       123
   macro avg     0.8954    0.8817    0.8809       123
weighted avg     0.8977    0.8862    0.8846       123



In [47]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.datasets import make_blobs
from sklearn.model_selection import GridSearchCV

X = df_changed.drop(["label"],axis=1)
y = df_changed.label

# 将数据集分为训练集和测试集Divide the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.6, random_state=123)

# 创建决策树分类器对象Create a decision tree classifier object
tree = DecisionTreeClassifier(random_state=123)

# 定义要进行交叉验证的参数候选值Define parameter candidates to be cross-validated
param_grid = {
    'max_depth': [2,3,5,7,9,10],
    'min_samples_split': [2,3,4,5,6,7,8,9,10],
    'min_samples_leaf': [2, 3,4,5],
    'criterion': ['gini', 'entropy']
}

# 定义交叉验证对象Define the cross-validation object
grid_search = GridSearchCV(tree, param_grid, cv=10)

# 执行交叉验证和参数调优Perform cross-validation and parameter tuning
grid_search.fit(X_train, y_train)

# 输出最佳模型的参数Output the parameters of the best model
print("Best Parameters:", grid_search.best_params_)

# 输出最佳模型的得分Output the parameters of the best model
print("Best Score:", grid_search.best_score_)



y_pred = grid_search.predict(X_test)

print("测试集分类报告Test Set Classification Report：\n")
# 输出分类报告
print(classification_report(y_test, y_pred,digits=4))





Best Parameters: {'criterion': 'gini', 'max_depth': 7, 'min_samples_leaf': 2, 'min_samples_split': 2}
Best Score: 0.9506172839506173
测试集分类报告Test Set Classification Report：

              precision    recall  f1-score   support

         0.0     1.0000    0.9744    0.9870        39
         1.0     1.0000    1.0000    1.0000        38
         2.0     0.9787    1.0000    0.9892        46

    accuracy                         0.9919       123
   macro avg     0.9929    0.9915    0.9921       123
weighted avg     0.9920    0.9919    0.9919       123



这里两组对比试验可以看到，未做交叉验证的模型，对于未知样本的预测准确率为：87.8%，相比训练集的100%有着一定的区别。  

相反，做了交叉验证和参数调优的决策树模型，对于未知样本的预测准确率为 99.19%，在交叉验证集合上面的准确率为96.25%，差距不是很大。  
(噪声会影响交叉验证集的结果)  

这说明，**参数调优和交叉验证在机器学习领域是非常必要的**。  
Here we can see that the model without cross-validation has a prediction accuracy of 87.8% for unknown samples, which is somewhat different from the 100% of the training set.  

On the contrary, the decision tree model with cross-validation and parameter tuning has a prediction accuracy of 99.19% for unknown samples and 96.25% on the cross-validation set. The gap is not very large.  
(Noise can affect the results of the cross-validation set)  

This shows that parameter tuning and cross-validation are very necessary in the field of machine learning.

### 3.4 Random forest

In [48]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_blobs
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

X = df_changed.drop(["label"],axis=1)
y = df_changed.label

# 将数据集分为训练集和测试集Divide the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=123)

rf = RandomForestClassifier(n_jobs=-1)
rf.fit(X_train, y_train)

y_pred = rf.predict(X_test)
report = classification_report(y_test, y_pred,digits=4)
print(report)



              precision    recall  f1-score   support

         0.0     1.0000    1.0000    1.0000        11
         1.0     1.0000    1.0000    1.0000        12
         2.0     1.0000    1.0000    1.0000        18

    accuracy                         1.0000        41
   macro avg     1.0000    1.0000    1.0000        41
weighted avg     1.0000    1.0000    1.0000        41



这里不难发现，随机森林这种森林系模型，相比决策树而言更不容易过拟合。  

接下来试一试减少训练样本数量，做同样的实验。  
It is not difficult to find here that the forest model of random forest is less prone to overfitting than decision tree.  

Next, try to reduce the number of training samples and do the same experiment.

### 3.5 Random forest(sample 40%)

In [49]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_blobs
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

X = df_changed.drop(["label"],axis=1)
y = df_changed.label

# 将数据集分为训练集和测试集Divide the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.7, random_state=123)

rf = RandomForestClassifier(n_jobs=-1)
rf.fit(X_train, y_train)

y_pred = rf.predict(X_test)
report = classification_report(y_test, y_pred,digits=4)
print(report)



              precision    recall  f1-score   support

         0.0     1.0000    0.9792    0.9895        48
         1.0     0.9535    1.0000    0.9762        41
         2.0     1.0000    0.9815    0.9907        54

    accuracy                         0.9860       143
   macro avg     0.9845    0.9869    0.9854       143
weighted avg     0.9867    0.9860    0.9861       143



这里会发现，尽管用30%的数据来训练模型，随机森林的表现依旧非常不错。  

接下来对其进行交叉验证和参数调优。  
Here you will find that despite using 30% of the data to train the model, the performance of the random forest is still very good.  

Next, perform cross-validation and parameter tuning on it.

In [50]:
reset -f

### 3.5 Random forest + cv (sample 40%)

In [51]:


import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import numpy as np
import matplotlib.pyplot as plt
import os
from sklearn.metrics import adjusted_rand_score
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report, roc_curve, auc
import numpy as np
import matplotlib.pyplot as plt
from itertools import combinations
# 引入必要的库
import numpy as np
import matplotlib.pyplot as plt
from itertools import cycle
from sklearn import svm, datasets
from sklearn.metrics import roc_curve, auc
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import label_binarize
from sklearn.multiclass import OneVsRestClassifier
from scipy import interp
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report, confusion_matrix, roc_curve, auc
import numpy as np
import matplotlib.pyplot as plt
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_blobs
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import classification_report

X = np.load('/home/mw/input/smiley2472/smiley_X.npy')
y = np.load('/home/mw/input/smiley2472/smiley_Y.npy')

import warnings 
X_changed=pd.DataFrame(X[0,:,:,0].reshape(-1,X.shape[1]*X.shape[1]))

for i in range(1,X.shape[0]):
    tmp = pd.DataFrame(X[i,:,:,0].reshape(-1,X.shape[1]*X.shape[1]))
    X_changed = X_changed.append(tmp)
    X_changed = X_changed.reset_index().drop(["index"],axis=1)

X_changed["label"]=y
df_changed = X_changed.copy()
df_changed.head()

from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_blobs
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

X = df_changed.drop(["label"],axis=1)
y = df_changed.label

# 将数据集分为训练集和测试集Divide the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.7, random_state=123)

# 创建决策树分类器对象Create a decision tree classifier object
rf = RandomForestClassifier(n_jobs=-1)

# 定义要调优的参数空间Define the parameter space to tune
param_grid = {
    'n_estimators': [50, 100, 200,400,500],
    'max_depth': [None, 1,2,3,4,5],
    'min_samples_leaf': [1, 2, 3,4],
    'min_samples_split': [2, 4,6, 8]
}

# 创建网格搜索对象Create a grid search object
grid_search = GridSearchCV(estimator=rf, param_grid=param_grid, cv=10)

# 在训练集上进行网格搜索Grid search on the training set
grid_search.fit(X_train, y_train)


GridSearchCV(cv=10, error_score='raise-deprecating',
             estimator=RandomForestClassifier(bootstrap=True, class_weight=None,
                                              criterion='gini', max_depth=None,
                                              max_features='auto',
                                              max_leaf_nodes=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              n_estimators='warn', n_jobs=-1,
                                              oob_score=False,
                                              random_state=None, verbose=0,
                                              warm_start=False),
             iid=

In [52]:

# 输出最佳参数组合Output the best parameter combination
print("cv Best Parameters: ", grid_search.best_params_)


# 输出最佳模型的得分# Output the score of the best model
print("cv Best Score:", grid_search.best_score_)


# 使用最佳参数的模型进行预测Use the model with the best parameters to make predictions
y_pred = grid_search.predict(X_test)

# 输出分类报告输出分类报告
report = classification_report(y_test, y_pred,digits=4)
print(report)



cv Best Parameters:  {'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 6, 'n_estimators': 50}
cv Best Score: 0.9836065573770492
              precision    recall  f1-score   support

         0.0     1.0000    0.8750    0.9333        48
         1.0     0.8542    1.0000    0.9213        41
         2.0     1.0000    0.9815    0.9907        54

    accuracy                         0.9510       143
   macro avg     0.9514    0.9522    0.9484       143
weighted avg     0.9582    0.9510    0.9515       143



可以发现，随机森林算法做交叉验证的结果和测试集的结果差距不大，这说明随机森林具有一定的抗过拟合能力。  

这是因为随机森林在构建决策树时引入了两种随机性：随机选择特征子集和随机选择样本子集。  

这些随机性可以减少决策树的方差，从而降低过拟合的风险。  

但是，在机器学习领域，交叉验证和模型调参还是非常有必要的，因为能够增加模型的泛化能力.  

It can be found that there is not much difference between the cross-validation results of the random forest algorithm and the results of the test set, which shows that the random forest has a certain ability to resist overfitting.  

This is because random forests introduce two kinds of randomness when building decision trees: random selection of feature subsets and random selection of sample subsets.  

These randomnesses reduce the variance of the decision tree, thereby reducing the risk of overfitting.  

However, in the field of machine learning, cross-validation and model tuning are still very necessary because they can increase the generalization ability of the model.